src/main/java/org/veupathdb/service/eda/ds/plugin/CategoricalCountsPlugin.java

```diff
@@ -0,0 +1,140 @@
+    return CategoricalCountsPostRequest.class;
+  }
+
+  @Override
+  protected Class<CategoricalCountsSpec> getVisualizationSpecClass() {
+    return CategoricalCountsSpec.class;
+  }
+
+  @Override
+  public ConstraintSpec getConstraintSpec() {
+    return new ConstraintSpec()
+      .dependencyOrder(List.of("variable"))
+      .pattern()
+        .element("variable")
+          .shapes(APIVariableDataShape.CATEGORICAL, APIVariableDataShape.ORDINAL)
+          .description("Variable must be categorical.")
+      .done();
+  }
+  
+  @Override
+  protected void validateVisualizationSpec(CategoricalCountsSpec pluginSpec) throws ValidationException {
+    validateInputs(new DataElementSet()
+      .var("variable", pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected List<StreamSpec> getRequestedStreams(CategoricalCountsSpec pluginSpec) {
+    String entityId = pluginSpec.getVariable().getEntityId();
+    _cacheKey = EncryptionUtil.md5(
+        JsonUtil.serializeObject(getSubsetFilters()) +
+        "|" + JsonUtil.serializeObject(pluginSpec)
+    );
+    _cachedResponse = RESULT_CACHE.get(_cacheKey);
+    if (_cachedResponse != null) LOG.info("Found cached result.");
+
+    return _cachedResponse != null ? Collections.emptyList() : ListBuilder.asList(
+      new StreamSpec(DEFAULT_SINGLE_STREAM_NAME, pluginSpec.getVariable().getEntityId())
+        .addVar(pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected void writeResults(OutputStream out, Map<String, InputStream> dataStreams) throws IOException {
+    CategoricalCountsSpec spec = getPluginSpec();
+
+    // create scanner and line parser
+    InputStreamReader isReader = new InputStreamReader(new BufferedInputStream(dataStreams.get(DEFAULT_SINGLE_STREAM_NAME)));
+    BufferedReader reader = new BufferedReader(isReader);
+    DelimitedDataParser parser = new DelimitedDataParser(reader.readLine(), TAB, true);
+
+    // establish column header indexes
+    Function<VariableSpec,Integer> indexOf = var -> parser.indexOfColumn(getUtil().toColNameOrEmpty(var)).orElseThrow();
+    int varIndex = indexOf.apply(spec.getVariable());
+
+    // this just needs to make a hashmap of values and their counts
+    Map<String, MutableInt> categoryCounts = new HashMap<>();
+    String nextLine = reader.readLine();
+
+    while (nextLine != null) {
+      String[] row = parser.parseLineToArray(nextLine);
+
+      categoryCounts.putIfAbsent(row[varIndex], new MutableInt());
+      categoryCounts.get(row[varIndex]).increment();
+
+      nextLine = reader.readLine();
+    }
+
+    // begin output object and single property containing array of elements
+    out.write("{\"categories\":[".getBytes(StandardCharsets.UTF_8));
+    boolean first = true;
+    for (String key : categoryCounts.keySet()) {
+      // write commas between elements
+      if (first) first = false; else out.write(',');
+      out.write(new JSONObject()
```

Comment:
I would say the only reason to do this kind of streaming is if you're taking values from a stream and don't want to store the entire JSON result in memory.  But since you're already storing the entire result in the HashMap above, it's probably no less safe to simply write:
```
out.write(new JSONObject().put("categories",
   categoryCounts.entrySet().stream()
      .map(e -> new JSONObject().put("binLabel", e.getKey()).put("value", e.getValue()))
      .toList()
).toString().getBytes(StandardCharsets.UTF_8));
```

Reply:
Not even sure you need the StandardCharsets.UTF_8 since that is probably the default.